### ISSUE TITLE
Tests Failing in FullDebug

### ISSUE THREAD

**Issue by RiccardoRossi** [KratosMultiphysics/Kratos#3254]
A number of tests fail or give deprecation warning when running in FullDebug (from the full suite)

also the test_KratosCore is now taking far too long. A naca0012 distance calculation test has been added and should be in another suite, not in the validation one. @rubenzorrilla @marcnunezc i think this is on you

**Comment by RiccardoRossi** [comment]
just to add that i also have

-D_GLIBCXX_DEBUG

defined as a compilation flag

**Comment by rubenzorrilla** [comment]
@rubenzorrilla I put it in there because it is a quite large test case. There are smaller similar test cases in the fast suite. @RiccardoRossi  Do you really want to move the naca one to the fast suite?

**Comment by RiccardoRossi** [comment]
@rubenzorrilla when i add -D_GLIBCXX_DEBUG it takes very long ... (minutes) 

actually my problem is that IT IS in the fast suite...i would put it in the nightly!

regarding the problem, i just rerun with clean master. FullDebug is ok, but when i add 
-D_GLIBCXX_DEBUG (extra debugging checks) i get the following error:

~~~
OK
Running tests for StructuralMechanicsApplication
/home/riccardo/Kratos/applications//StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
/usr/include/c++/7/debug/safe_iterator.h:149:
Error: attempt to copy-construct an iterator from a singular iterator.

Objects involved in the operation:
    iterator "this" @ 0x0x7f8df00251f0 {
      type = __gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::pair<Kratos::VariableData const*, void*>*, std::__cxx1998::vector<std::pair<Kratos::VariableData const*, void*>, std::allocator<std::pair<Kratos::VariableData const*, void*> > > >, std::__debug::vector<std::pair<Kratos::VariableData const*, void*>, std::allocator<std::pair<Kratos::VariableData const*, void*> > > > (mutable iterator);
      state = singular;
    }
    iterator "other" @ 0x0x7f8df0025270 {
      type = __gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::pair<Kratos::VariableData const*, void*>*, std::__cxx1998::vector<std::pair<Kratos::VariableData const*, void*>, std::allocator<std::pair<Kratos::VariableData const*, void*> > > >, std::__debug::vector<std::pair<Kratos::VariableData const*, void*>, std::allocator<std::pair<Kratos::VariableData const*, void*> > > > (mutable iterator);
      state = singular;
      references sequence with type 'std::__debug::vector<std::pair<Kratos::VariableData const*, void*>, std::allocator<std::pair<Kratos::VariableData const*, void*> > >' @ 0x0x55e442e1a060
    }
Running cpp tests
Importing    KratosFSIApplication 
Initializing KratosFSIApplication... 
Importing    KratosExternalSolversApplication 
Initializing KratosExternalSolversApplication... 
Importing    KratosMetisApplication 
Initializing Kratos MetisApplication... 
Importing    KratosConvectionDiffusionApplication 
 KRATOS ___ ___  _  ___   __   ___ ___ ___ ___ 
       / __/ _ || || | | / /__|   |_ _| __| __|
      | (_| (_) | .` || V /___| |) | || _|| _| 
       |___|___/|_||_| |_/    |___/___|_| |_|  APPLICATION
Importing    KratosFluidDynamicsApplication 
Initializing KratosFluidDynamicsApplication... 
Importing    KratosMeshMovingApplication 
KRATOS __  __        _    __  __         _                       
      |  \/  |___ __| |_ |  \/  |_____ _(_)_ _  __ _           
      | |\/| / -_|_-< ' \| |\/| / _ \ V / | ' \/ _` |       
      |_|  |_\___/__/_||_|_|  |_\___/\_/|_|_||_\__, |        
                                               |___/  application
Initializing KratosMeshMovingApplication...                      
Importing    KratosStructuralMechanicsApplication 
     KRATOS   ___|  |                   |                   |                     
            \___ \  __|  __| |   |  __| __| |   |  __| _` | |                   
                  | |   |    |   | (    |   |   | |   (   | |                     
            _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS     
/usr/include/c++/7/debug/safe_iterator.h:374:
Error: attempt to advance a singular iterator 0 steps, which falls outside 
its valid range.

Objects involved in the operation:
    iterator @ 0x0x7ffd4829f300 {
      type = __gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<__gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >*, std::__cxx1998::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::__debug::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >*, std::__cxx1998::vector<__gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >*, std::__cxx1998::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::__debug::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::allocator<__gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >*, std::__cxx1998::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::__debug::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > > > > >, std::__debug::vector<__gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >*, std::__cxx1998::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::__debug::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::allocator<__gnu_debug::_Safe_iterator<__gnu_cxx::__normal_iterator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >*, std::__cxx1998::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > >, std::__debug::vector<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > >, std::allocator<std::shared_ptr<Kratos::Node<3ul, Kratos::Dof<double> > > > > > > > > (mutable iterator);
      state = singular;
    }
Aborted (core dumped)
~~~

**Comment by rubenzorrilla** [comment]
@RiccardoRossi ok! I thought that it was already in the validation suite. I have no objections to migrate it.

**Comment by philbucher** [comment]
@RiccardoRossi Please wrap code or output in `~~~` (I updated your post) otherwise it is very hard to read

Also I am confused as to which tests exactly are failing now, can you please specify?

**Comment by RiccardoRossi** [comment]
@philbucher to be honest i am also confused of which test is actually failing... i simply know that one is failing (with -D_GLIBCXX_DEBUG)

**Comment by philbucher** [comment]
I also ran the tests now with -D_GLIBCXX_DEBUG and for me the following cpp-test fails:
TestDynamicBinsNearestPointAborted
~~~
/usr/include/c++/5/debug/safe_iterator.h:408:error: attempt to advance a 
    singular iterator 0 steps, which falls outside its valid range.

Objects involved in the operation:
iterator @ 0x0x7ffe26076ef0 {
type = N11__gnu_debug14_Safe_iteratorIN9__gnu_cxx17__normal_iteratorIPNS0_INS2_IPSt10shared_ptrIN6Kratos4NodeILm3ENS4_3DofIdEEEEENSt9__cxx19986vectorIS9_SaIS9_EEEEENSt7__debug6vectorIS9_SD_EEEENSC_ISJ_SaISJ_EEEEENSH_ISJ_SL_EEEE (mutable iterator);
  state = singular;
}
~~~

@roigcarlo I think this is your test

**Comment by roigcarlo** [comment]
Yep that one is mine. Looking to it thx

**Comment by loumalouomega** [comment]
Is this still happening?

**Comment by philbucher** [comment]
yes, ping @roigcarlo

**Comment by roigcarlo** [comment]
Should be solved in #4325 as well
